distributor.go

```diff
@@ -189,30 +193,45 @@ func metricNameFromLabelMatchers(matchers ...*metric.LabelMatcher) (model.LabelV
 func (d *Distributor) Query(ctx context.Context, from, to model.Time, matchers ...*metric.LabelMatcher) (model.Matrix, error) {
 	var result model.Matrix
 	err := instrument.TimeRequestHistogram("duration", d.queryDuration, func() error {
-		metricName, err := metricNameFromLabelMatchers(matchers...)
-		if err != nil {
-			return err
-		}
+		fpToSampleStream := map[model.Fingerprint]*model.SampleStream{}
+
+		// Fetch samples from all ingesters and group them by fingerprint (unsorted
+		// and with overlap).
+		for _, c := range d.queryRing.GetAll() {
+			client, err := d.getClientFor(c.Hostname)
+			if err != nil {
+				return err
+			}
+			matrix, err := client.Query(ctx, from, to, matchers...)
+			if err != nil {
+				return err
+			}
 
-		userID, err := user.GetID(ctx)
-		if err != nil {
-			return err
+			for _, ss := range matrix {
+				fp := ss.Metric.Fingerprint()
+				if mss, ok := fpToSampleStream[fp]; !ok {
+					fpToSampleStream[fp] = &model.SampleStream{
+						Metric: ss.Metric,
+						Values: ss.Values,
+					}
+				} else {
+					mss.Values = append(fpToSampleStream[fp].Values, ss.Values...)
+				}
+			}
 		}
 
-		collector, err := d.ring.Get(tokenFor(userID, metricName))
-		if err != nil {
-			return err
+		// Sort and dedupe samples.
+		for _, ss := range fpToSampleStream {
+			sortable := timeSortableSamplePairs(ss.Values)
+			sort.Sort(sortable)
+			// TODO: Dedupe samples. Not strictly necessary.
 		}
 
-		client, err := d.getClientFor(collector.Hostname)
-		if err != nil {
-			return err
+		matrix := make(model.Matrix, 0, len(fpToSampleStream))
+		for _, ss := range fpToSampleStream {
+			matrix = append(matrix, ss)
```

Comment:
Thanks, this was meant to append to the `result` defined at the top. Pushing fix.